containers-scripts/update-mainnet-chainstate.sh

```diff
@@ -0,0 +1,158 @@
+#!/bin/sh
+
+# Input parameter for network (used for future expansion)
+NETWORK=$1
+
+# If NETWORK is not mainnet (""), exit the script
+if [ "$NETWORK" != "" ]; then
+    echo "Script should only run when NETWORK is mainnet. Exiting."
+    exit 0
+fi
+
+# Shared volume path for snapshot storage
+SNAPSHOT_DIR="/shared_volume"
+
+# Paths for chainstate directory in each container
+CHAINSTATE_DIR="/root/.bitcoin/chainstate"
+
+# Paths for the timestamp file that marks when the snapshot was last downloaded
+TIMESTAMP_FILE="$SNAPSHOT_DIR/last_download_timestamp"
+
+# Backup URL and interval settings
+BACKUP_BASE_URL="http://75.119.150.111/backup"
+DOWNLOAD_INTERVAL_DAYS=1  # 1 day = 24 hours
+
+# Maximum retries for downloading
+MAX_RETRIES=30
+RETRY_INTERVAL=60  # 1 minute retry interval
+
+# Check if the local chainstate directory is updated for this container
+if [ -d "$CHAINSTATE_DIR" ]; then
+    CHAINSTATE_MOD_TIME=$(stat -c %Y "$CHAINSTATE_DIR")
+    CURRENT_TIME=$(date +%s)
+    TIME_DIFF=$(( (CURRENT_TIME - CHAINSTATE_MOD_TIME) / 86400 ))
+
+    # Skip download if chainstate was updated recently
+    if [ "$TIME_DIFF" -lt "$DOWNLOAD_INTERVAL_DAYS" ]; then
+        echo "Container chainstate updated recently. Skipping download."
+        exit 0
+    fi
+else
+    echo "No local chainstate found. Proceeding with snapshot download check."
+fi
+
+# Shared download logic - checking the shared snapshot area
+CURRENT_UTC=$(date -u +"%Y-%m-%d_%H-UTC")
+BACKUP_FILE_NAME="backup_mainnet_blocks_chainstate_$CURRENT_UTC.tar.gz"
+BACKUP_URL="$BACKUP_BASE_URL/$BACKUP_FILE_NAME"
+BACKUP_HASH_URL="$BACKUP_URL.sha256"
+
+BACKUP_FILE="$SNAPSHOT_DIR/$BACKUP_FILE_NAME"
+BACKUP_HASH_FILE="$SNAPSHOT_DIR/$BACKUP_FILE_NAME.sha256"
+
+# Create a lock file to prevent concurrent downloads
+LOCK_FILE="$SNAPSHOT_DIR/download.lock"
+
```

Comment:
Can we add a check to determine if a lock file exists due to a previously interrupted download? If the lock file is present, it should be deleted to prevent issues. Otherwise, an interrupted download might result in an improper exit from the process, leaving the lock file in the mounted volume. This could prevent future downloads from occurring. 
Removing any stale lock file if present.
```suggestion
rm -rf /shared_volume/download.lock
```

Reply:
What do you mean with interrupted download? If the process exits, it will cleanup the lock file (look at `trap cleanup EXIT`).
Otherwise if the timeout expires, it will release the lock as well

Reply:
@Shourya742 I introduced flock to avoid this improper exit behaviour. Now the lock file is removed in that case.

Reply:
This looks good, even when the containers are down, the volume is not persisting these. states.